net/rpc/server.go

```diff
@@ -191,13 +191,34 @@ type Server struct {
 	freeReq    *Request
 	respLock   sync.Mutex // protects freeResp
 	freeResp   *Response
+
+	serverInterceptor ServerInterceptor
 }
 
 // NewServer returns a new Server.
 func NewServer() *Server {
 	return &Server{}
 }
 
+// NewServerWithOpts returns a new Server with the following functional options.
+func NewServerWithOpts(options... func(*Server)) *Server {
+	s := &Server{}
+	for _, option := range options {
+		option(s)
+	}
+	return s
+}
+
+func WithServerInterceptor(interceptor ServerInterceptor) func(*Server) {
+	return func(s *Server) {
+		s.serverInterceptor = interceptor
+	}
+}
+
+// ServerInterceptor acts a middleware hook on the server side of the RPC call. The interceptor must
+// invoke the handler argument for the RPC request to continue.
+type ServerInterceptor func(serviceMethod string, argv, replyv reflect.Value, handler func())
```

Comment:
It's possible we could have other interceptors for the server in the future. What do you think about calling this `ServiceCallInterceptor`. I believe `service.call` is the operation we are intercepting.

```suggestion
type ServiceCallInterceptor func(serviceMethod string, argv, replyv reflect.Value, handler func())
```

Reply:
Hmm

> have other interceptors for the server in the future. 

Ideally, if we have multiple `ServiceCallInterceptor`s, we probably want to do some chaining, how it's done on gRPC. Or mathematically, F(G(..(H(X))).

I've not thought at all about intercepting other operations -- 🤔 so renaming this makes 100% sense then

---

I'm still partial to seeing `ServerXInterceptor` so that, in the future, if we want to have a `ClientXInterceptor` we can slot that in as well. But I can make this addition in an upcoming commit here if it sounds good?